conans/client/graph/graph.py

```diff
@@ -83,7 +83,7 @@ def __init__(self, ref, conanfile, context, recipe=None, path=None):
         # on this node. It includes regular (not private and not build requires) dependencies
         self._transitive_closure = OrderedDict()
         self.inverse_closure = set()  # set of nodes that have this one in their public
-        self.ancestors = None  # set{ref.name}
+        self._ancestors = _NodeOrderedDict()  # set{ref.name}
```

Comment:
It doesn't create an infinite loop. The key is `(name, context)`, as soon as a recipe creates a loop using `build_requires` it will repeat `(name, "build")` and Conan will raise an ERROR saying there is a loop.